src/test/testFrontierDistance.cpp

```diff
@@ -0,0 +1,180 @@
+    nobMilitary* milBld1 =
+      dynamic_cast<nobMilitary*>(BuildingFactory::CreateBuilding(world, BLD_WATCHTOWER, milBld1Pos, 1, NAT_BABYLONIANS));
+
+    unsigned distance0 = milBld0->GetFrontierDistance();
+    unsigned distance1 = milBld1->GetFrontierDistance();
+
+    BOOST_REQUIRE_EQUAL(distance0, 3u);
+    BOOST_REQUIRE_EQUAL(distance0, distance1);
+}
+
+BOOST_FIXTURE_TEST_CASE(FrontierDistanceMiddle, FrontierWorldMiddle)
+{
+    MapPoint milBld0Pos = MapPoint(1, 20);
+    nobMilitary* milBld0 =
+      dynamic_cast<nobMilitary*>(BuildingFactory::CreateBuilding(world, BLD_WATCHTOWER, milBld0Pos, 0, NAT_BABYLONIANS));
+
+    MapPoint milBld1Pos = MapPoint(31, 20);
+    nobMilitary* milBld1 =
+      dynamic_cast<nobMilitary*>(BuildingFactory::CreateBuilding(world, BLD_WATCHTOWER, milBld1Pos, 1, NAT_BABYLONIANS));
+
+    unsigned distance0 = milBld0->GetFrontierDistance();
+    unsigned distance1 = milBld1->GetFrontierDistance();
+
+    BOOST_REQUIRE_EQUAL(distance0, 1u);
+    BOOST_REQUIRE_EQUAL(distance0, distance1);
+}
+
+BOOST_FIXTURE_TEST_CASE(FrontierDistanceInland, FrontierWorldBig)
+{
+    MapPoint milBld0Pos = MapPoint(1, 60);
+    nobMilitary* milBld0 =
+      dynamic_cast<nobMilitary*>(BuildingFactory::CreateBuilding(world, BLD_WATCHTOWER, milBld0Pos, 0, NAT_BABYLONIANS));
+
+    MapPoint milBld1Pos = MapPoint(61, 60);
+    nobMilitary* milBld1 =
+      dynamic_cast<nobMilitary*>(BuildingFactory::CreateBuilding(world, BLD_WATCHTOWER, milBld1Pos, 1, NAT_BABYLONIANS));
+
+    unsigned distance0 = milBld0->GetFrontierDistance();
+    unsigned distance1 = milBld1->GetFrontierDistance();
+
+    BOOST_REQUIRE_EQUAL(distance0, 0u);
+    BOOST_REQUIRE_EQUAL(distance0, distance1);
+}
+
+BOOST_FIXTURE_TEST_CASE(FrontierDistanceNearFrontierAddonActive, FrontierWorldSmall)
+{
+    DescIdx<TerrainDesc> tWater(0);
+    for(; tWater.value < world.GetDescription().terrain.size(); tWater.value++)
+    {
+        if(world.GetDescription().get(tWater).kind == TerrainKind::WATER && !world.GetDescription().get(tWater).Is(ETerrain::Walkable))
+            break;
+    }
+    this->ggs.setSelection(AddonId::FRONTIER_DISTANCE_REACHABLE, 1);
+    for(int i = 1; i < world.GetHeight(); i++)
+    {
+        world.GetNodeWriteable(MapPoint(5, i)).t1 = tWater;
+    }
+
+    MapPoint milBld0Pos = MapPoint(1, 10);
+    nobMilitary* milBld0 =
+      dynamic_cast<nobMilitary*>(BuildingFactory::CreateBuilding(world, BLD_WATCHTOWER, milBld0Pos, 0, NAT_BABYLONIANS));
+
+    MapPoint milBld1Pos = MapPoint(10, 10);
+    nobMilitary* milBld1 =
+      dynamic_cast<nobMilitary*>(BuildingFactory::CreateBuilding(world, BLD_WATCHTOWER, milBld1Pos, 1, NAT_BABYLONIANS));
+
+    unsigned distance0 = milBld0->GetFrontierDistance();
+    unsigned distance1 = milBld1->GetFrontierDistance();
+
+    BOOST_REQUIRE_EQUAL(distance0, 0u);
+    BOOST_REQUIRE_EQUAL(distance0, distance1);
+}
+
+BOOST_FIXTURE_TEST_CASE(FrontierDistanceMiddleAddonActive, FrontierWorldMiddle)
```

Comment:
In the active test you use `FrontierWorldMiddle` but in the other you use `FrontierWorldSmall`. Maybe you could unify the 2 test cases to ensure this is the same? (Run in a loop, once with addon=1, once with addon=0 but keep the world setup the same. You could either issue a recalculation of the frontier distances or destroy and rebuild the buildings)

**Important**: Change the test case first before fixing the (possible) bug I mentioned above. IMO 1 test case MUST fail due to that bug. If not, either it is not a bug or you need another test case that triggers the bug.

Reply:
Okay i think i have managed it. there were multiple problems... The biggest was, that the water is set in my test case was walkable, so the whole test case was wrong... After i had fixed that i realized that the Condition check did not really  work for the test case.
I have fixed both problems and now i think it should work as expected.

Reply:
Now you are using unreachable terrain? Why not use the unwalkable water (like in some cases before)? That would be more common than lava. (or even test both)

What exactly was wrong with the condition check? IMO it was correct: You have to have all terrains around a point "not unreachable". Otherwise you could land/walk next to lava...

BTW: For your world structs you could use typedefs. They are shorter :)

Reply:
i have changed the test back to the "land part" of PathConditionHuman.
I have tried to unit test with water, but i have to be honest, i have no idea why its not working if i change it to water. for me it seems that water is threaded as walkable (maybe 1-2 fields??) because in real game scenario it works for me...)

Reply:
Found your problem. You put a single colon of water there. Just draw the triangles on a piece of paper and you will see why this fails. My blueprint (large so one can reuse it): 
![grid](https://user-images.githubusercontent.com/309017/35183504-f5b6c802-fde7-11e7-86d2-c6c3b695af88.png)

And with 1 column of water on t1 only (might be actually t2 but doesn't matter for showing): 
![water](https://user-images.githubusercontent.com/309017/35183591-13b3c282-fde9-11e7-9520-1012b682cb8b.jpg)


As you can easily see, walking is very much possible. You'd need to block it completely by making it at least 2 triangles, so one would have to walk ON water to cross. (Walking next to water is allowed)

Note: There is an edge case I just remembered: If there is a road, people can walk on water. A road is possible when every node has a) no danger terrain around and b) any terrain is (at least) walkable. This means that something like this should count as connected: 
![water2](https://user-images.githubusercontent.com/309017/35183593-19d1da5a-fde9-11e7-8a1d-a19f4a88cafa.jpg)
To handle this let the checker return true for any transition and only check nodes.